File: src/utils/env.ts
/**
 * Environment Variable Utilities
 *
 * Trace:
 *   spec_id: SPEC-GHACTION-001
 *   task_id: TASK-GHACTION-001
 */

/**
 * Required environment variable keys
 */
const REQUIRED_ENV_KEYS = ['USER_ID', 'PASSWORD', 'TELEGRAM_BOT_TOKEN', 'TELEGRAM_CHAT_ID'] as const;

export type RequiredEnvKey = (typeof REQUIRED_ENV_KEYS)[number];

/**
 * Get a required environment variable
 * Throws an error if the variable is not set
 */
export function getEnv(key: RequiredEnvKey): string {
  const value = process.env[key];
  if (!value) {
    throw new Error(`Missing required environment variable: ${key}`);
  }
  return value;
}

/**
 * Validate all required environment variables
 * Throws an error with all missing variables if any are not set
 */
export function validateEnv(): void {
  const missing = REQUIRED_ENV_KEYS.filter((key) => !process.env[key]);

  if (missing.length > 0) {
    throw new Error(`Missing required environment variables: ${missing.join(', ')}`);
  }
}

File: internal/env/env.go
// Package env is the sole boundary for reading required environment variables.
// Domain code must never read os.Getenv directly.
package env

import (
	"fmt"
	"os"
	"strings"
)

// Required lists the environment variables that must be present at runtime.
var Required = []string{
	"USER_ID",
	"PASSWORD",
	"TELEGRAM_BOT_TOKEN",
	"TELEGRAM_CHAT_ID",
}

// Get returns the value of a required environment variable, or an error if it
// is unset or empty.
func Get(key string) (string, error) {
	v := os.Getenv(key)
	if v == "" {
		return "", fmt.Errorf("missing required environment variable: %s", key)
	}
	return v, nil
}

// Validate ensures all required environment variables are present.
func Validate() error {
	var missing []string
	for _, key := range Required {
		if os.Getenv(key) == "" {
			missing = append(missing, key)
		}
	}
	if len(missing) > 0 {
		return fmt.Errorf("missing required environment variables: %s", strings.Join(missing, ", "))
	}
	return nil
}
